pkg/quotes/index.go

```diff
@@ -0,0 +1,178 @@
+package quotes
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/shopspring/decimal"
+)
+
+type PriceCache interface {
+	GetEMA(market string) (decimal.Decimal, int64)
+	UpdateEMA(market string, newValue decimal.Decimal)
+}
+
+type IndexAggregator struct {
+	weightsURL string
+	weightsMap map[DriverType]decimal.Decimal
+	prices     PriceCache
+	drivers    []Driver
+	aggregated chan TradeEvent
+	outbox     chan<- TradeEvent
+}
+
+func NewIndex(driverConfigs []Config, remoteWeights string, outbox chan<- TradeEvent) *IndexAggregator {
+	aggregated := make(chan TradeEvent, 128)
+	drivers := []Driver{}
+	for _, d := range driverConfigs {
+		driver, err := NewDriver(d, aggregated)
+		if err != nil {
+			logger.Info("cannot start a driver: ", err)
+			return nil
+		}
+
+		drivers = append(drivers, driver)
+	}
+
+	return &IndexAggregator{
+		prices:     NewPricesCache(),
+		drivers:    drivers,
+		outbox:     outbox,
+		aggregated: aggregated,
+	}
+}
+
+func (a *IndexAggregator) Start(markets []Market) error {
+	go func() {
+		for {
+			err := a.fetchWeights(context.Background())
+			if err != nil {
+				logger.Error(err)
+			}
+			<-time.After(5 * time.Minute)
+		}
+	}()
+
+	for _, d := range a.drivers {
+		go func(d Driver) {
+			logger.Info("starting index quotes service")
+
+			markets := make([]Market, 0, len(markets))
+			for _, m := range markets {
+				market := Market{BaseUnit: m.BaseUnit, QuoteUnit: m.QuoteUnit}
+				markets = append(markets, market)
+			}
+
+			if err := d.Start(markets); err != nil {
+				logger.Warn(err.Error())
+			}
+		}(d)
+	}
+	go func() {
+		for event := range a.aggregated {
+			indexPrice := a.indexPrice(event)
+			a.outbox <- indexPrice
+		}
+	}()
+	return nil
+}
+
+func (a *IndexAggregator) Subscribe(m Market) error {
+	for _, d := range a.drivers {
+		err := d.Subscribe(m)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *IndexAggregator) Stop() error {
+	for _, d := range a.drivers {
+		err := d.Stop()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *IndexAggregator) indexPrice(event TradeEvent) TradeEvent {
+	lastEMA, timestamp := a.prices.GetEMA(event.Market)
+	weight := a.weightsMap[event.Source]
+
+	newEMA := EMA20(lastEMA, event.Price.Mul(event.Amount.Mul(weight).Div(a.activeWeights())))
+	if time.Now().Unix() >= timestamp+60 {
+		a.prices.UpdateEMA(event.Market, newEMA)
+	}
+
+	event.Price = newEMA
+	event.Source = DriverIndex
+
+	return event
+}
+
+func (a *IndexAggregator) activeWeights() decimal.Decimal {
+	total := decimal.Zero
+	for _, w := range a.weightsMap {
+		total = total.Add(w)
+	}
+	return total
+}
+
+func (a *IndexAggregator) fetchWeights(ctx context.Context) error {
+	req, err := http.NewRequest(http.MethodGet, a.weightsURL, nil)
```

Comment:
lets define in the pkg, without external source, you can keep this function for just in case for future